### PR TITLE
Aruco javascript fix and added functionality

### PR DESCRIPTION
**Pull request by urbste**
### Pull Request Readiness Checklist

resolves https://github.com/opencv/opencv_contrib/issues/2526
resolves https://github.com/opencv/opencv/issues/15514

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ x] I agree to contribute to the project under Apache 2 License.
- [ x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders_only=docs,Custom
build_image:Docs=docs-js
build_image:Custom=javascript
buildworker:Custom=linux-4
```

### CODE REVIEW

**Comment by urbste** [comment]
#15514

**Comment by urbste** [comment]
Corresponding PR in https://github.com/opencv/opencv_contrib/pull/2769

File: modules/js/src/embindgen.py [hunk 1]
```diff
@@ -635,7 +635,7 @@ def gen_function_binding(self, func, class_info):
                 if ptr_type in type_dict:
                     ret_type = type_dict[ptr_type]
             for key in type_dict:
-                if key in ret_type:
+                if key in ret_type and "DetectorParameters" not in ret_type:
```
Comment (dkurt) [inline modules/js/src/embindgen.py:638]: Can you please describe why DetectorParameters cannot be parsed with common code?

Comment (urbste) [reply modules/js/src/embindgen.py:638]: As mentioned in the issue the  .replace operator is used to replace parts of the string with strings from a type_dict. During this operation the Param string from DetectorParameters is replaced with **cv::Param**, leading then to the error .constructor(select_overload<Ptr<**Detectorcv::Parameters**>()> in the binding.cpp.

Is that what you meant?

Comment (dkurt) [reply modules/js/src/embindgen.py:638]: Yeah, exactly! Do you know where exactly in code generation script there is such pattern matching? In my point of view, it's better to fix `cv::Param` replacement rather than add an excetion for `DetectorParameters`

Comment (urbste) [reply modules/js/src/embindgen.py:638]: You are right. It's not a nice way because for future bindings this could happen again. I can have a look this afternoon. However, I have not understood the binding generation completely as the code is not that well documented and very long. I just debugged myself to the error. So if anyone with more experience with the binding generation has an idea it would be great.

Comment (urbste) [reply modules/js/src/embindgen.py:638]: So I had a quick look at it. I don't see a quick way to make this more elegant from my side, sorry. Maybe someone else has an idea?!

File: platforms/js/build_js.py [hunk 2]
```diff
@@ -239,6 +245,8 @@ def build_loader(self):
     # Write a path to modify file like argument of this flag
     parser.add_argument('--config', default=os.path.join(os.path.dirname(os.path.abspath(__file__)), 'opencv_js.config.py'),
                         help="Specify configuration file with own list of exported into JS functions")
+    parser.add_argument('--opencv_contrib_dir', default='',
+                        help="Path to /opencv_contrib/modules folder. Can be used to build Aruco marker support into opencvjs.")
```
Comment (dkurt) [inline platforms/js/build_js.py:249]: Not needed. Users should specify it by `--cmake_option="-DOPENCV_EXTRA_MODULES_PATH=opencv_contrib/modules"`

Comment (dkurt) [reply platforms/js/build_js.py:249]: Let's add this to the doc instead: https://docs.opencv.org/master/d4/da1/tutorial_js_setup.html

File: modules/js/src/embindgen.py (hunk 1 above)
Comment (dkurt) [reply modules/js/src/embindgen.py:638]: Seems I found a problem. Please use the following construction to manage types replacement more accurate:
```patch
--- a/modules/js/src/embindgen.py
+++ b/modules/js/src/embindgen.py
@@ -636,7 +636,10 @@ class JSWrapperGenerator(object):
                     ret_type = type_dict[ptr_type]
             for key in type_dict:
                 if key in ret_type:
-                    ret_type = ret_type.replace(key, type_dict[key])
+                    # Replace types. Instead of ret_type.replace we use regular
+                    # expression to exclude false matches.
+                    # See https://github.com/opencv/opencv/issues/15514
+                    ret_type = re.sub('(^|[^\w])' + key + '($|[^\w])', type_dict[key], ret_type)
```
in our case, works extra replacement from `Param` to `cv::Param` for `Ptr<DetectorParameters>`. Proposed patch applies replacement not just by pattern but pattern with non-alphanumeric symbols before and after.

File: platforms/js/opencv_js.config.py [hunk 3]
```diff
@@ -74,9 +74,12 @@
         'aruco_Board': ['create'],
         'aruco_GridBoard': ['create', 'draw'],
         'aruco_CharucoBoard': ['create', 'draw'],
+        'aruco_DetectorParameters': ['create']
         }
 
-calib3d = {'': ['findHomography', 'calibrateCameraExtended', 'drawFrameAxes', 'estimateAffine2D', 'getDefaultNewCameraMatrix', 'initUndistortRectifyMap', 'Rodrigues']}
+calib3d = {'': ['findHomography', 'calibrateCameraExtended', 'drawFrameAxes', 'estimateAffine2D', \
+                'getDefaultNewCameraMatrix', 'initUndistortRectifyMap', 'Rodrigues', \
+                'solvePnP', 'solvePnPRansac', 'solvePnPRefineLM']}
```
Comment (dkurt) [inline platforms/js/opencv_js.config.py:82]: Seems unrelated to this PR

**Comment by dkurt** [comment]
Hi, @urbste, can you please resolve merge conflicts and consider comments? Thansk!

File: modules/js/src/embindgen.py (hunk 1 above)
Comment (urbste) [reply modules/js/src/embindgen.py:638]: Awesome! Updated it.

File: platforms/js/build_js.py (hunk 2 above)
Comment (urbste) [reply platforms/js/build_js.py:249]: I updated the documentation and removed the argument from the parser.

File: platforms/js/opencv_js.config.py (hunk 3 above)
Comment (urbste) [reply platforms/js/opencv_js.config.py:82]: It is not a bugfix that's true. However if we enable Aruco markers in the build, we usually would want to use the Aruco board for camera calibration, pose estimation and so on (that's what I am using it for), so I thought it would be straightforward to just add the solvePnP methods to the js build. I could remove it if you want, however, I don't see any major problems in just adding the 3 functions?!

File: platforms/js/build_js.py [hunk 4]
```diff
@@ -140,6 +140,7 @@ def get_cmake_cmd(self):
                "-DBUILD_PACKAGE=OFF",
                "-DBUILD_TESTS=OFF",
                "-DBUILD_PERF_TESTS=OFF"]
+
```
Comment (dkurt) [inline platforms/js/build_js.py:143]: Please remove extra changes from this file

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 5]
```diff
@@ -127,6 +127,13 @@ Building OpenCV.js from Source
     python ./platforms/js/build_js.py build_js --build_test
     @endcode
 
+-#  [optional] To build support for Aruco markers from the OpenCV contrib modules append `--cmake_option="-DOPENCV_EXTRA_MODULES_PATH=/path/to/opencv_contrib/modules/"`
```
Comment (dkurt) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown:130]: No need to highlight only Aruco. Just "To add contrib modules, add a flag"

**Comment by dkurt** [comment]
Please resolve merge conflicts - rebase with the latest master branch

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown (hunk 5 above)
Comment (urbste) [reply doc/js_tutorials/js_setup/js_setup/js_setup.markdown:130]: Wow you are quick  ^^

Comment (urbste) [reply doc/js_tutorials/js_setup/js_setup/js_setup.markdown:130]: Done

File: platforms/js/build_js.py (hunk 4 above)
Comment (urbste) [reply platforms/js/build_js.py]: Done

File: platforms/js/build_js.py [hunk 6]
```diff
@@ -184,6 +184,7 @@ def get_build_flags(self):
             flags += "-msimd128 "
         if self.options.build_flags:
             flags += self.options.build_flags
+
```
Comment (dkurt) [inline platforms/js/build_js.py:187]: extra new line

File: doc/js_tutorials/js_setup/js_setup/js_setup.markdown [hunk 7]
```diff
@@ -127,6 +127,13 @@ Building OpenCV.js from Source
     python ./platforms/js/build_js.py build_js --build_test
     @endcode
 
+-#  [optional] To build support for OpenCV contrib modules append `--cmake_option="-DOPENCV_EXTRA_MODULES_PATH=/path/to/opencv_contrib/modules/"`
```
Comment (dkurt) [inline doc/js_tutorials/js_setup/js_setup/js_setup.markdown:130]: "To enable OpenCV contrib modules append .." sounds more correct, as for me.

Comment (urbste) [reply doc/js_tutorials/js_setup/js_setup/js_setup.markdown]: Ok

**Comment by urbste** [comment]
How do I resolve the merge conflicts. Sorry I am not that fit with git rebase from a remote? I did 
``` bash 
git fetch origin 
git rebase origin/master
```
However nothing really happened. And locally I do not get the merge conflicts. What am I doing wrong?

**Comment by dkurt** [comment]
Try the following:
1. Make sure that `origin` is `opencv/opencv`:
```
$ git remote -v
```
2. Update
```
git checkout master
git pull origin HEAD
```
3. Rebase
```
git rebase master feature/aruco_js_fix
```
4. Resolve conflict in file
5. `git add filename` + `git rebase --continue`
6. `git push fork feature/aruco_js_fix -f`

**Comment by urbste** [comment]
Thanks for the tips. Still got problems. I did:
``` bash
git remote -v
origin	https://github.com/urbste/opencv (fetch)
origin	https://github.com/urbste/opencv (push)
upstream	https://github.com/opencv/opencv (fetch)
upstream	https://github.com/opencv/opencv (push)
```
origin was not opencv/opencv. So I did
```bash
git remote set-url origin https://github.com/opencv/opencv
git checkout master
git pull origin HEAD
git rebase master feature/aruco_js_fix
git mergetool
git rebase --continue

```
However, I only had one conflict in the documentation file, not the emdingen.py as indicated here on github. So something is still wrong I guess... :(

This is not working:
```bash
git push fork feature/aruco_js_fix -f
fatal: 'fork' does not appear to be a git repository
```

Sorry for the inconveniece. :/

**Comment by urbste** [comment]
Ok I think I found a solution. I did not rebase but merge the master into my branch.

**Review by dkurt — APPROVED** [review]
:+1: Well done!

**Comment by dkurt** [comment]
@urbste, instead of fork you had to write your fork name. Anyway, now it OK.

**Comment by urbste** [comment]
@dkurt I do have some more significant changes in the Aruco module, that we would also like to share with the community. I still have forked the repositories in my github account. However I would like to transfer the repositories to my companies github team. Do you know if this PR will be influenced by that?

**Comment by dkurt** [comment]
@urbste, aruco module is located in a separate repository. `OPENCV_EXTRA_MODULES_PATH` can with with any custom path to modules.

**Comment by urbste** [comment]
I think this issue is also related to this PR: https://github.com/opencv/opencv_contrib/issues/2526
